promotions/app/models/solidus_promotions/promotion_code/batch_builder.rb

```diff
@@ -0,0 +1,72 @@
+# frozen_string_literal: true
+
+module SolidusPromotions
+  class PromotionCode < Spree::Base
+    class BatchBuilder
+      attr_reader :promotion_code_batch, :options
+
+      delegate :promotion, :number_of_codes, :base_code, to: :promotion_code_batch
+
+      DEFAULT_OPTIONS = {
+        random_code_length: 6,
+        batch_size: 1000,
+        sample_characters: ("a".."z").to_a + (2..9).to_a.map(&:to_s)
+      }.freeze
+
+      def initialize(promotion_code_batch, options = {})
+        @promotion_code_batch = promotion_code_batch
+        options.assert_valid_keys(*DEFAULT_OPTIONS.keys)
+        @options = DEFAULT_OPTIONS.merge(options)
+      end
+
+      def build_promotion_codes
+        generate_random_codes
+        promotion_code_batch.update!(state: "completed")
+      rescue StandardError => e
+        promotion_code_batch.update!(
+          error: e.inspect,
+          state: "failed"
+        )
+        raise e
+      end
+
+      private
+
+      def generate_random_codes
+        created_codes = promotion_code_batch.promotion_codes.count
+
+        batch_size = @options[:batch_size]
+
+        while created_codes < number_of_codes
+          max_codes_to_generate = [batch_size, number_of_codes - created_codes].min
+
+          new_codes = Array.new(max_codes_to_generate) { generate_random_code }.uniq
+          codes_for_current_batch = get_unique_codes(new_codes)
+
+          codes_for_current_batch = codes_for_current_batch.map do |value|
+            SolidusPromotions::PromotionCode.create!(
+              value: value,
+              promotion: promotion,
+              promotion_code_batch: promotion_code_batch
+            )
+          rescue ActiveRecord::RecordInvalid
```

Comment:
@jarednorman and @tvdeyen had a concern about the memory usage [here](https://github.com/solidusio/solidus/pull/5383#discussion_r1391519251).

Reply:
Thanks for the heads-up! I've added a commit implementing @jarednorman s suggestion, reverting to the performance characteristics of the legacy promotion batch builder. It's still not great, see https://github.com/solidusio/solidus/pull/5805/commits/64f18c1af5c9f6850a7b45cef9733c4bcdc50a9b - however, this is an example implementation that people can override, and whoever has 5 million promotion codes to create can surely afford to optimize this. The batch builder class is configurable.